system/security/pkencrypt/pkencrypt.cpp

```diff
@@ -0,0 +1,260 @@
+#include <openssl/err.h>
+
+//Helper to public/private key encrytption
+static CriticalSection encrypterCrit;//serializes encryption
+static CriticalSection decrypterCrit;//serializes decryption
+
+
+class _CPKEncryptionManager : implements IPKEncryptionManager
+{
+private:
+#ifdef _USE_ZLIB
+    StringAttr   privateKeyFile;
+    StringAttr   publicKeyFile;
+    StringBuffer privateKey;
+    IZDecryptor* pDecrypter;//single instance of decrypter, which is not reentrant
+    IZEncryptor* pEncrypter;//single instance of encrypter, which is not reentrant
+#endif
+    bool         encrpytConfigured;
+    bool         decrpytConfigured;
+
+public:
+    _CPKEncryptionManager()
+    {
+        encrpytConfigured = false;
+        decrpytConfigured = false;
+        pEncrypter = nullptr;
+        pDecrypter = nullptr;
+#ifdef _USE_ZLIB
+        //query private key file location from environment.conf
+        StringBuffer configFileSpec;
+        configFileSpec.set(CONFIG_DIR).append(PATHSEPSTR).append("environment.conf");
+        Owned<IProperties> conf = createProperties(configFileSpec.str(), true);
+        if (conf)
+        {
+            publicKeyFile.set(conf->queryProp("HPCCPublicKey"));
+            privateKeyFile.set(conf->queryProp("HPCCPrivateKey"));
+            encrpytConfigured = !publicKeyFile.isEmpty();
+            decrpytConfigured = !privateKeyFile.isEmpty();
+        }
+#else
+        WARNLOG("PKEncrypt: Platform built without ZLIB!");
+#endif
+    }
+
+    virtual ~_CPKEncryptionManager()
+    {
+#ifdef _USE_ZLIB
+        {
+            CriticalBlock b(encrypterCrit);
+            if (pEncrypter)
+                releaseIZ(pEncrypter);
+        }
+
+        {
+            CriticalBlock b(decrypterCrit);
+            if (pDecrypter)
+                releaseIZ(pDecrypter);
+        }
+#endif
+    }
+
+    virtual inline bool isEncrypterConfigured()
+    {
+        return encrpytConfigured;
+    }
+
+    virtual bool PKEncrypt(const char * _str, StringAttr & _key, StringAttr & _token)
+    {
+        if (!encrpytConfigured)
+            throw MakeStringException(-1, "PKEncrypt error: encryption not configured");
+
+        if (_str == nullptr  || *_str == '\0')
+            return false;
+
+#ifdef _USE_ZLIB
+        CriticalBlock b(encrypterCrit);
+
+        if (!pEncrypter)
+        {
+            StringBuffer pubKey;
+            pubKey.loadFile(publicKeyFile);//throws on error
+            pEncrypter = createZEncryptor(pubKey.str());
+            if (pEncrypter)
+            {
+                pEncrypter->setEncoding(true);
+#ifdef _DEBUG
+                pEncrypter->setTraceLevel(100);
+#endif
+            }
+            else
+                throw MakeStringException(-1, "PKEncrypt error: creating PKI encrypter using key file %s", publicKeyFile.str());
+        }
+
+        IZBuffer* buf = nullptr;
+        IZBuffer* key = nullptr;
+        int rc = pEncrypter->encrypt(strlen(_str), (unsigned char*)_str, key, buf);
```

Comment:
Do we always require _str arg to be null terminated ?

Reply:
Yes